indico/modules/rb_new/client/js/modules/admin/SettingsPage.jsx

```diff
@@ -0,0 +1,220 @@
+import {Form, Header, Message, Placeholder} from 'semantic-ui-react';
+import {PluralTranslate, Translate} from 'indico/react/i18n';
+import {
+    formatters, getChangedValues, FieldCondition, ReduxFormField, ReduxCheckboxField,
+    validators as v
+} from 'indico/react/forms';
+import * as adminActions from './actions';
+import * as adminSelectors from './selectors';
+import CategoryList from './CategoryList';
+
+
+class SettingsPage extends React.PureComponent {
+    static propTypes = {
+        settingsLoaded: PropTypes.bool.isRequired,
+        settings: PropTypes.object.isRequired,
+        actions: PropTypes.exact({
+            fetchSettings: PropTypes.func.isRequired,
+            updateSettings: PropTypes.func.isRequired,
+        }).isRequired,
+    };
+
+    componentDidMount() {
+        const {actions: {fetchSettings}} = this.props;
+        fetchSettings();
+    }
+
+    handleSubmit = async (data, form) => {
+        const {actions: {updateSettings}} = this.props;
+        const changedValues = getChangedValues(data, form);
+        const rv = await updateSettings(changedValues);
+        if (rv.error) {
+            return rv.error;
+        }
+    };
+
+    render() {
+        const {settingsLoaded, settings} = this.props;
+        if (!settingsLoaded) {
+            return (
+                <Placeholder>
+                    <Placeholder.Line />
+                    <Placeholder.Line />
+                    <Placeholder.Line />
+                    <Placeholder.Line />
+                    <Placeholder.Line />
+                </Placeholder>
+            );
+        }
+
+        return (
+            <>
+                <Header as="h2">
+                    <Translate>General Settings</Translate>
+                </Header>
+                <FinalForm onSubmit={this.handleSubmit}
+                           initialValues={settings}
+                           initialValuesEqual={_.isEqual}
+                           subscription={{
+                               submitting: true,
+                               hasValidationErrors: true,
+                               pristine: true,
+                               submitSucceeded: true,
+                               dirtySinceLastSubmit: true,
+                           }}>
+                    {fprops => (
+                        <Form onSubmit={fprops.handleSubmit}
+                              success={fprops.submitSucceeded && !fprops.dirtySinceLastSubmit}>
+                            <Field name="tileserver_url" component={ReduxFormField} as="input"
+                                   format={formatters.trim} formatOnBlur
+                                   label={Translate.string('Tileserver URL')}
+                                   disabled={fprops.submitting}
+                                   validate={val => {
+                                       if (!val) {
+                                           return undefined;
```

Comment:
Could just return but I guess being explicit here is nice.

Reply:
yes, i don't like mixing bare returns and returns with a value in the same function.